File: php/src/Simple/Composability/WinningOrder.php
<?php

namespace ImmutableDatamodel\Simple\Composability;

use DateTimeImmutable;
use ImmutableDatamodel\Shared\Address;
use ImmutableDatamodel\Shared\OrderConstraint;
use ImmutableDatamodel\Simple\Deliverable;
use ImmutableDatamodel\Simple\DeliveringOrder;
use ImmutableDatamodel\Simple\Order;

readonly class WinningOrder implements Order, Deliverable
{
    /**
     * @param string $orderId
     * @param OrderConstraint[] $constraints
     * @param string $productSerialNo
     */
    public function __construct(
        private string $orderId,
        private array  $constraints,
        private string $productSerialNo
    ) {
    }

    public function orderId(): string
    {
        return $this->orderId;
    }

    public function constraints(): array
    {
        return $this->constraints;
    }

    public function productSerialNo(): string
    {
        return $this->productSerialNo;
    }

    public function deliver(Address $deliveryAddress, DateTimeImmutable $deliveryDate): DeliveringOrder
    {
        return new DeliveringOrder($this->orderId(), $this->constraints(), $deliveryAddress, $deliveryDate);
    }
}

File: golang/src/simple/winning_order.go
package simple

import (
	"example/shared"
	"time"
)

type WinningOrder struct {
	orderId         string
	constraints     []shared.OrderConstraint
	productSerialNo string
}

func (wo WinningOrder) OrderId() string {
	return wo.orderId
}

func (wo WinningOrder) Constraints() []shared.OrderConstraint {
	return wo.constraints
}

func (wo WinningOrder) Deliver(deliveryAddress shared.Address, deliveryDate time.Time) (DeliveringOrder, error) {
	return NewDeliveringOrder(wo, deliveryAddress, deliveryDate)
}
